transform_stats_to_metrics.go

```diff
@@ -0,0 +1,276 @@
+	if vd == nil || len(vd.Rows) == 0 {
+		return nil, nil
+	}
+
+	// Given that view.Data only contains Start, End
+	// the timestamps for all the row data will be the exact same
+	// per aggregation. However, the values will differ.
+	// Each row has its own tags.
+	startTimestamp := timeToProtoTimestamp(vd.Start)
+	endTimestamp := timeToProtoTimestamp(vd.End)
+
+	timeseries := make([]*metricspb.TimeSeries, 0, len(vd.Rows))
+	// It is imperative that the ordering of "LabelValues" matches those
+	// of the Label keys in the metric descriptor.
+	for _, row := range vd.Rows {
+		labelValues := labelValuesFromTags(row.Tags)
+		point := rowToPoint(row, endTimestamp)
+		timeseries = append(timeseries, &metricspb.TimeSeries{
+			StartTimestamp: startTimestamp,
+			LabelValues:    labelValues,
+			Points:         []*metricspb.Point{point},
+		})
+	}
+
+	if len(timeseries) == 0 {
+		return nil, nil
+	}
+
+	return timeseries, nil
+}
+
+func timeToProtoTimestamp(t time.Time) *timestamp.Timestamp {
+	unixNano := t.UnixNano()
+	return &timestamp.Timestamp{
+		Seconds: int64(unixNano / 1e9),
+		Nanos:   int32(unixNano % 1e9),
+	}
+}
+
+func rowToPoint(row *view.Row, endTimestamp *timestamp.Timestamp) *metricspb.Point {
+	pt := &metricspb.Point{
+		Timestamp: endTimestamp,
+	}
+
+	switch data := row.Data.(type) {
+	case *view.CountData:
+		pt.Value = &metricspb.Point_Int64Value{Int64Value: data.Value}
+
+	case *view.DistributionData:
+		pt.Value = &metricspb.Point_DistributionValue{
+			DistributionValue: &metricspb.DistributionValue{
+				Count:   data.Count,
+				Sum:     float64(data.Count) * data.Mean, // because Mean := Sum/Count
+				Buckets: exemplarsToDistributionBuckets(data.ExemplarsPerBucket),
+
+				SumOfSquaredDeviation: data.SumOfSquaredDev,
+			}}
+
+	case *view.LastValueData:
+		pt.Value = &metricspb.Point_DoubleValue{DoubleValue: data.Value}
+
+	case *view.SumData:
+		pt.Value = &metricspb.Point_DoubleValue{DoubleValue: data.Value}
+	}
+
+	return pt
+}
+
+func exemplarsToDistributionBuckets(exemplars []*exemplar.Exemplar) []*metricspb.DistributionValue_Bucket {
+	if len(exemplars) == 0 {
+		return nil
+	}
+
+	distBuckets := make([]*metricspb.DistributionValue_Bucket, 0, len(exemplars))
+	for _, exmplr := range exemplars {
+		if exmplr == nil {
+			continue
+		}
+
+		distBuckets = append(distBuckets, &metricspb.DistributionValue_Bucket{
+			Count: 1, // TODO: (@odeke-em) examine if OpenCensus-Go stores the count of values in the bucket
+			Exemplar: &metricspb.DistributionValue_Exemplar{
+				Value:       exmplr.Value,
+				Timestamp:   timeToTimestamp(exmplr.Timestamp),
+				Attachments: exmplr.Attachments,
+			},
+		})
+	}
+
+	return distBuckets
+}
+
+func labelValuesFromTags(tags []tag.Tag) []*metricspb.LabelValue {
+	if len(tags) == 0 {
+		return nil
+	}
+
+	labelValues := make([]*metricspb.LabelValue, 0, len(tags))
+	for _, tag_ := range tags {
+		labelValues = append(labelValues, &metricspb.LabelValue{
+			Value: tag_.Value,
```

Comment:
Just to confirm: in Go `tag.Value` will always be non-nil? Is there a case that we should set `HasValue` to false?

Reply:
@songy23 I am not sure about this but is makes more sense to use HasValue if Value != ""